MaxText/maxtext_utils.py

```diff
@@ -139,6 +139,11 @@ def calculate_tflops_training_per_device(config, log=True):
       attention_flops * config.num_decoder_layers * 3 / 10**12
   )
 
+  # in gemma2 we combine [local attention, global attention] into one decoder block, so need to multiply the
```

Comment:
Don't we also need to multiply the FF by 2 as well since there are 2 FF per gemma2 layer? I'm not sure the best way to handle this, I would rather not have decoder-block specific logic, however here it is necessary. Can we move it to its own subfunction, e.g. "if config.decoer_Block == gemma2: return calculate_gemma2_flops(config)
else:
calculate standard_flops(config)"

If you can think of a better solution let me know

Reply:
I think we should have calculate_gemma2_flops defined inside of the gemma2 decoder block definition actually, and call it from max_utils. Ideally some inheritance with a default method inherited that works for most models, but can be overriden, we max_utils will call decoder_block.calculate_tflops()...

We can do in the inheritance later, but for this PR we should at least calculate the flops accurately ideally in its own method in either max_utils or the gemma2 decoder block

Reply:
Thanks for the catch! while refactoring I found since most of the code in the new `calculate_gemma2_flops` are the same I use a local variable `num_decoder_layers` and multiply it by 2 when using the `gemma2` decoder block, please let me know if this works, just updated the branch on this as well, thanks again!

Reply:
doesn't local/sliding attention use less flops than global attention? I think thats the main reason gemma2 uses sliding attention?

Reply:
yeah haven't factored that in since we are using `dot_product` which still computes the full attention, will update to  `calculate_gemma2_flops` to include the sliding window attention flops calculation, thanks for the comment!

Reply:
calculate_flops is meant to calculate the necessary flops for the model to train, not the amount of flops it actually uses. This allows it to be used to measure MFU

Reply:
Thanks again for the details! updated the calculation to include sliding window attention for gemma2 and factor in the 2 stacked decoder layer in ffn layers